gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.extension
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.backendsapi.BackendsApiManager
+
+import org.apache.spark.sql.catalyst.expressions.aggregate.{Complete, Final, Partial}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.aggregate.{BaseAggregateExec, HashAggregateExec, ObjectHashAggregateExec}
+
+/**
+ * Merge two phase hash-based aggregate into one aggregate in the spark plan:
+ *
+ * Examples:
+ *
+ * HashAggregate(t1.i, SUM, final) \| => HashAggregate(t1.i, SUM, complete) HashAggregate(t1.i, SUM,
+ * partial)
+ */
+case class MergeTwoPhasesHashBaseAggregate() extends Rule[SparkPlan] {
+
+  private def isPartialAgg(partialAgg: BaseAggregateExec, finalAgg: BaseAggregateExec): Boolean = {
+    if (
+      partialAgg.aggregateExpressions.forall(x => x.mode == Partial) &&
+      finalAgg.aggregateExpressions.forall(x => x.mode == Final)
+    ) {
+      (finalAgg.logicalLink, partialAgg.logicalLink) match {
+        case (Some(agg1), Some(agg2)) => agg1.sameResult(agg2)
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
```

Comment:
It is better to use `PhysicalPlanSelector.maybe` to check if guten is enabled.

Reply:
Perhaps you should check whether the `SparkPlan` has been tagged as `TRANSFORM_UNSUPPORTED`. There are some rules earlier that will tag the `SparkPlan`, for example, `FallbackOnANSIMode` will tag all `SparkPlan`s as `TRANSFORM_UNSUPPORTED` when ANSI mode is enabled. In this case, we can also avoid merging the aggregations.

Reply:
done